### ISSUE TITLE
Remove cruft from platform descriptions

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nixpkgs#34274]
Simply put, I should be able to do
```
$ nix-build '<nixpkgs>' --arg crossSystem '{ config = "aarch64-unknown-linux-gnu"; }' -A stdenv
```
and that should work. I could make it work right now with a dirty white-list and fall-backs, but that's kind of defeats the point.

Currently, the biggest culprits are
 1. Random top-level fields, like `withTLS` and `openssl.system`. See https://github.com/NixOS/nixpkgs/blob/master/lib/systems/examples.nix
 2. The `platform` field, see https://github.com/NixOS/nixpkgs/blob/master/lib/systems/platforms.nix 
 3. `arch` which overlaps with the far more principled `parsed,arch`. See `elaborate` https://github.com/NixOS/nixpkgs/blob/master/lib/systems/default.nix#L17.

The general design issue is we need to handle the per-package random crap in a more modular way, without giving up the possibility of the user to override with their own values. My GNU ld emulation picking logic at https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/bintools-wrapper/default.nix#L167-L179 is certainly in a more modular location, but also cannot be nicely overridden.

#### Checklist

 - [x] #34303: `withTLS`
 - [x] #34315: `bigEndian`, `openssl.system`
 - [x] #34351: linux kernel cleanups

-------

CC @dtzWill @dezgeg @Infinisil @teto

**Comment by teto** [comment]
I have a PR in the works (should be ready tomorrow) that allows to do
my_kernel = buildLinux {
kernelAutoModules=true, ignoreConfigErrors=true, kernelPreferBuiltin
} which makes for a much better experience as it looks more like your typical derivation.
 I let the parameters kernelAutoModules etc as defaults in the platform sets. Do you suggest I remove these parameters from these platform sets https://github.com/NixOS/nixpkgs/blob/master/lib/systems/platforms.nix#L36 ?

**Comment by Ericson2314** [comment]
@teto I'd say only leave things for kernel headers there, as that actually effects everything (on Linux).

**Comment by dezgeg** [comment]
I agree for refactoring `openssl` and most of the kernel stuff out, but much of the other stuff seems quite hard to get rid of.

**Comment by dezgeg** [comment]
I made some WIP stuff for the kernel one time: https://github.com/NixOS/nixpkgs/compare/master...dezgeg:kernel-plat-fix?expand=1, but IIRC some of the kernels failed to boot during runtime, still need to debug that one day.

**Comment by Ericson2314** [comment]
Per https://github.com/NixOS/nixpkgs/pull/34645#discussion_r166325878 @dtzWill and I need to figure out whether to moved `libc` into `parsed`, or similar. @dezgeg or anyone else have any opinions? https://github.com/NixOS/nixpkgs/pull/34444 is also impacted.

**Comment by dtzWill** [comment]
For what it's worth, I've started using just the "config" in my cross specifications and so far so good!

I don't cross-build kernels as part of my work or testing, and I note that is the remaining TODO item :).  Just wanted to chime in that the rest is in a good place!

Minor convenience thing: it's annoying to specify `--arg crossSystem '{ config = "x"; }'` instead of something like `--argstr crossConfig x`, especially as the good work on this issue makes anything other than config unnecessary (at least usually).  Is there a simple way we could allow the shorter syntax without compromising abstractions?
(similar to how we handle "system" today, minus special nix support)

**Comment by Ericson2314** [comment]
@dtzWill Maybe we can get `--argstr crossSystem.config` to work for 1.12? Though I feel bad asking for new features for that. I'm hesitant to add any more sugar as the platform stuff is already far too complicated.

**Comment by dtzWill** [comment]
@Ericson2314 thanks for the issue, and agreed not priority for nix update.  I was just wanting to discuss it to see what you had in mind.

I'll follow that issue, neat idea!

I agree we shouldn't further complicate the platform stuff, haha, it's actually partly why I asked O:).

**Comment by Ericson2314** [comment]
I made a bunch of changes so this is now true. All the examples contain at most `config`, `platform`, and additional platform-specific flags like `useIosPrebuilt` that are new and don't overlap. Yay!

**Comment by Profpatsch** [comment]
This is still linked in the manual, should be removed if closed.

**Comment by siraben** [comment]
Still important to me

**Comment by siraben** [comment]
I can take a look at this once #105364 is merged.

**Comment by siraben** [comment]
@Ericson2314 @Profpatsch Is this still an issue? Can the reference in the manual be deleted?

**Comment by Ericson2314** [comment]
If we can merge https://github.com/NixOS/nixpkgs/pull/107214, and (better still) rearrange `examples.nix` and `platforms.nix` into something more sensible afterwords, then I'd say this is done.

**Comment by Ericson2314** [comment]
Added the checkboxes for those.

**Comment by Ericson2314** [comment]
@siraben OK, just the final coding step (and manual victor lap) remains. If you want to take a stab at this, that would be much appreciated!

**Comment by sambacha** [comment]
This issue is still referenced in the manual BTW
